### PR TITLE
core/vm: Hide read only flag from Interpreter interface

### PR DESCRIPTION
**Pull request by chfast**
Not really a functional change, but makes Interface interface a bit more stateless and abstract.

### CODE REVIEW

**Comment by chfast** [comment]
Done.

File: core/vm/interpreter.go [hunk 1]
```diff
@@ -138,6 +134,11 @@ func (in *EVMInterpreter) Run(contract *Contract, input []byte) (ret []byte, err
 	in.evm.depth++
 	defer func() { in.evm.depth-- }()
 
+	if readOnly && !in.readOnly {
```
Comment (holiman) [inline core/vm/interpreter.go]: Please add comment that was in the previous code:
```
	// Make sure the readonly is only set if we aren't in readonly yet
	// this makes also sure that the readonly flag isn't removed for
	// child calls.
```

**Comment by holiman** [comment]
I can live with this, but it forces the global `readonly`:ness to be part of the internal state of whatever evm is used anyway, so I can't imagine it changes a lot for the evm implementation.

**Comment by chfast** [comment]
> I can live with this, but it forces the global readonly:ness to be part of the internal state of whatever evm is used anyway, so I can't imagine it changes a lot for the evm implementation.

Obviously this change is dictated by EVMC design. The EVMC tries to keep the responsibility for EVM features totally inside the VMs, if feasible. This makes VM "stateless" because VM does not need to pass any information between executions, all information is included in parameters of the execute function.

So in the EVMC PR for geth this flag is never used by the EVMC VM and is also not accurate, so querying it might give wrong results.

I'm not sure this provided enough arguments for it to make you feel comfortable with the change.

File: core/vm/interpreter.go (hunk 1 above)
Comment (chfast) [reply core/vm/interpreter.go]: The comment has been added back.

**Pull request merged**